Add support for smartthings SAMSUNG_VD_AMBIENT capability

daviddob opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

### Background
Ambient mode is available for Samsung TVs including most notably "The Frame", but has also come to other Samsung QLED models. [Samsung KB Article](https://www.samsung.com/us/support/answer/ANS00082422/)

Along with it came the addition of the `samsungvd.ambient` capability with the `setAmbientOn` command

This functions as a toggle, with each api call either turning ambient mode on or off.

Relevant API Calls and smartthings-cli command for testing
```
SMARTTHINGS_DEBUG=true smartthings devices:commands xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx 'samsungvd.ambient:setAmbientOn'

[2025-04-05T18:21:40.209] [DEBUG] rest-client - making axios request: {"url":"https://api.smartthings.com/devices/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/commands","method":"post","headers":{"Content-Type":"application/json;charset=utf-8","Accept":"application/json","User-Agent":"@smartthings/cli/1.10.3 wsl-x64 node-v18.5.0","Accept-Language":"C","Authorization":"Bearer 31d1ccb3-xxxx-xxxx-xxxx-xxxxxxxxxxxx"},"data":{"commands":[{"component":"main","capability":"samsungvd.ambient","command":"setAmbientOn","arguments":[]}]}}
Command executed successfully
```

### Implemented Changes

- Adds a new button entity for all devices supporting SAMSUNG_VD_AMBIENT capability.
- Button executes the SAMSUNG_VD_AMBIENT capability with the command SET_AMBIENT_ON. This is implemented as a button as the api is a toggle and does not have support for reading the status/eventing if the TV is in ambient mode.

- Tested and working using 2024 Frame TV


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

<details>

<summary>Test Pass Report </summary>

```
---------- coverage: platform linux, python 3.13.2-final-0 -----------
Name                                                              Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------------
homeassistant/components/smartthings/__init__.py                    197     12    94%   125-129, 161-162, 196-197, 313, 363, 367, 377, 406
homeassistant/components/smartthings/application_credentials.py      26      7    73%   51-57
homeassistant/components/smartthings/binary_sensor.py                56      0   100%
homeassistant/components/smartthings/button.py                       25      0   100%
homeassistant/components/smartthings/climate.py                     220      0   100%
homeassistant/components/smartthings/config_flow.py                  47      0   100%
homeassistant/components/smartthings/const.py                        17      0   100%
homeassistant/components/smartthings/cover.py                        60      2    97%   90, 150
homeassistant/components/smartthings/diagnostics.py                  25      1    96%   44
homeassistant/components/smartthings/entity.py                       40      0   100%
homeassistant/components/smartthings/event.py                        25      0   100%
homeassistant/components/smartthings/fan.py                          58      0   100%
homeassistant/components/smartthings/light.py                        95      2    98%   96, 175
homeassistant/components/smartthings/lock.py                         34      1    97%   78
homeassistant/components/smartthings/media_player.py                127      0   100%
homeassistant/components/smartthings/number.py                       33      0   100%
homeassistant/components/smartthings/scene.py                        21      0   100%
homeassistant/components/smartthings/select.py                       40      0   100%
homeassistant/components/smartthings/sensor.py                       90      1    99%   1144
homeassistant/components/smartthings/switch.py                       63      0   100%
homeassistant/components/smartthings/update.py                       35      0   100%
homeassistant/components/smartthings/util.py                         30      1    97%   29
homeassistant/components/smartthings/valve.py                        26      0   100%
-----------------------------------------------------------------------------------------------
TOTAL                                                              1390     27    98%

------------------------------------------------------------------------------------------ snapshot report summary -------------------------------------------------------------------------------------------
609 snapshots passed.

Results (374.38s (0:06:14)):
    1073 passed
```

</details>    
    



If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository] - N/A

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. - N/A

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

daviddob committed:
Add support for smartthings SAMSUNG_VD_AMBIENT capability

This adds support for the SAMSUNG_VD_AMBIENT capability with
the command SET_AMBIENT_ON. This is implemented as a button
as the api is a toggle and does not have support for reading
the status/eventing if the TV is in ambient mode.

Tested and working using 2024 Frame TV

daviddob committed:
Merge branch 'dev' into smartthings-samsung-ambientmode

joostlek commented on homeassistant/components/smartthings/strings.json:
```diff
@@ -62,6 +62,9 @@
         "name": "Reset water filter"
       },
+      "toggle_tv_ambient_mode": {
+        "name": "Toggle TV Ambient Mode"
```
Please have this in sentence case

joostlek reviewed (changes requested):
Let's also add an icon, and let's make sure we have a device fixture for this. Also, we should test this with other users. 

Also, do we also have a turn off command?

daviddob commented:
> Let's also add an icon

Do you have a specific icon in mind? Could use a TV of some sort like `mdi:television-shimmer` or `mdi:remote-tv` as it's somewhat of a TV remote function with the newer Samsung TV remotes having a dedicated button for ambient mode.

> let's make sure we have a device fixture for this. 

I'll see what I can look into here - I located the button test suite, will try to figure out how it works and update the tests accordingly.

> Also, we should test this with other users.

Absolutely - I'll see if I can link folks running into this lack of feature (ambient mode) to this PR for testing - I know it's requested on a bunch of the HACS repos for Samsung tizen tvs. Is there an easy way for someone to test out a patch to a core component without cloning a fork and setting up a full custom_components override?

> Also, do we also have a turn off command?

This is a toggle, even though the command is named `setAmbientOn`, calling it again when ambient mode is currently on turns ambient mode off. There is no corresponding `setAmbientOff` or similar commands.

joostlek commented:
Do we know if ambient is running?

daviddob commented:
I've looked through all of the available capabilities+ sensors and while there are attributes for `samsungvd.ambientContent` and `samsungvd.ambient`, none of them are populated or update with any information at the moment in smartthings.

In terms of detection if it is in ambient mode, `tvChannel:tvChannelName` gets set to `art` when it is in ambient mode, however I believe this is afflicted by the polling issue that's been discussed in other threads. If I have the smart things app open, HA updates that field immediately, but if not it sometimes lags behind getting the update so could be finicky?

Could possibly look to move this to a switch that uses the `tvChannelName` to sense state, especially if progress will be made on the polling vs eventing with smarthings as a whole, but feels a bit brittle/prone to confusing behavior with current implementation. 

Thoughts?

joostlek commented:
If we can detect this, ideally we create a switch out of it. I will try to check with SmartThings if there's an update on this

daviddob committed:
Fix case + add default icon

daviddob committed:
Merge branch 'smartthings-samsung-ambientmode' of github.com:daviddob/homeassistant-core into smartthings-samsung-ambientmode

daviddob committed:
Add test fixture + device json for frame tv 2024

daviddob committed:
Merge branch 'home-assistant:dev' into smartthings-samsung-ambientmode

daviddob commented:
Addressed all of the above comments:
- Sentence case
- Icon
- Testing Fixtures

I included a smartthings api response for the 2024 Samsung Frame TV device and status for use in the fixture - could also likely be added to pysmartthings test fixtures for future reference of a device that includes ambient mode support.

Any updates from smartthings on whether we can get an endpoint/event to detect status such that we can consider converting from a button to a switch?

I can try implementing it with the `tvChannelName == 'art'` status, though it has issues w.r.t sync/refresh as is discussed on discord so would rather avoid it until there's a recommended path forward to make that more responsive/reliable.
